keps/sig-auth/20190711-external-credential-providers.md

```diff
@@ -0,0 +1,243 @@
+
+There are 2 modes of authentication:
+
+1. bearer tokens
+1. mTLS
+
+Provider response is cached and reused in future requests.
+
+Client is configured with a binary path, optional arguments and environment
+variables to pass to it.
+
+### Provider configuration
+
+Configuration is provided via users section of `kubeconfig` file:
+
+```
+apiVersion: v1
+kind: Config
+users:
+- name: my-user
+  user:
+    exec:
+      # API version to use when decoding the ExecCredentials resource. Required.
+      apiVersion: "client.authentication.k8s.io/v1beta1"
+
+      # Command to execute. Required.
+      command: "example-client-go-exec-plugin"
+
+      # Arguments to pass when executing the plugin. Optional.
+      args:
+      - "arg1"
+      - "arg2"
+
+      # Environment variables to set when executing the plugin. Optional.
+      env:
+      - name: "FOO"
+        value: "bar"
+clusters:
+- name: my-cluster
+  cluster:
+    server: "https://1.2.3.4:8080"
+    certificate-authority: "/etc/kubernetes/ca.pem"
+contexts:
+- name: my-cluster
+  context:
+    cluster: my-cluster
+    user: my-user
+current-context: my-cluster
+```
+
+`apiVersion` specifies the expected version of this API that the plugin
+implements. If the version differs, client must return an error.
+
+`command` specifies the path to the provider binary. The file at this path must
+be readable and executable by the client process.
+
+`args` specifies extra arguments passed to the executable.
+
+`env` specifies environment variables to pass to the provider. The environment
+variables set in the client process are not passed.
+
+### Provider input format
+
+```
+{
+  "apiVersion": "client.authentication.k8s.io/v1",
+  "kind": "ExecCredential"
+}
+```
+
+Provider can safely ignore `stdin` since input object doesn't carry any data.
+
+### Provider output format
+
+```
+{
+  "apiVersion": "client.authentication.k8s.io/v1",
+  "kind": "ExecCredential",
+  "status": {
+    "expirationTimestamp": "$EXPIRATION",
+    "token": "$BEARER_TOKEN",
+    "clientKeyData": "$CLIENT_PRIVATE_KEY",
+    "clientCertificateData": "$CLIENT_CERTIFICATE",
+  }
+}
+```
+
+`EXPIRATION` contains the RFC3339 timestamp with credential expiry. This field
+is not set in `handshaker` mode. Client can cache provided credentials until
+this time.
+
+After `EXPIRATION`, client must execute the provider again for any new
+connections. For `client_key` mode, this applies even if returned certificate
+is still valid.
+
+`BEARER_TOKEN` contains a token for use in `Authorization` header of HTTP
+requests.
+
+`CLIENT_PRIVATE_KEY` and `CLIENT_CERTIFICATE` contain client TLS credentials in
+PEM format. The certificate must be valid at the time of execution. These
+credentials are used for mTLS handshakes.
+
+### Risks and Mitigations
+
+#### Client authentication to the binary
+
+Credential provider binary does not do any authentication/authorization of the
+caller. Caller of the binary does not pass any credentials either.
```

Comment:
other than envvars or arguments specified in the kubeconfig file

Reply:
Right, updated